cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/dataset/input/BasicPartitionedFileSetInputContext.java

```diff
@@ -42,34 +49,80 @@
 
   private static final Type STRING_PARTITION_KEY_MAP_TYPE = new TypeToken<Map<String, PartitionKey>>() { }.getType();
 
-  private final Path inputPath;
-  private final String mappingString;
+  private final Map<String, PartitionKey> pathToPartitionMapping;
 
-  private PartitionKey partitionKey;
+  private final boolean isCombineInputFormat;
+  private final Configuration conf;
+
+  private final Supplier<Path[]> inputPaths;
+  private List<PartitionKey> partitionKeys;
+
+  // for caching in case of CombineFileInputFormat
+  private String currentInputfileName;
+  private PartitionKey currentPartitionKey;
 
   BasicPartitionedFileSetInputContext(MultiInputTaggedSplit multiInputTaggedSplit) {
     super(multiInputTaggedSplit.getName());
 
     InputSplit inputSplit = multiInputTaggedSplit.getInputSplit();
-    if (!(inputSplit instanceof FileSplit)) {
-      throw new IllegalArgumentException(String.format("Expected a '%s', but got '%s'.",
-                                                       FileSplit.class.getName(), inputSplit.getClass().getName()));
+    if (inputSplit instanceof FileSplit) {
+      isCombineInputFormat = false;
+      Path path = ((FileSplit) inputSplit).getPath();
+      inputPaths = Suppliers.ofInstance(new Path[] { path });
+    } else if (inputSplit instanceof CombineFileSplit) {
+      isCombineInputFormat = true;
+      inputPaths = Suppliers.ofInstance(((CombineFileSplit) inputSplit).getPaths());
+    } else {
+      throw new IllegalArgumentException(String.format("Expected either a '%s' or a '%s', but got '%s'.",
+                                                       FileSplit.class.getName(), CombineFileSplit.class.getName(),
+                                                       inputSplit.getClass().getName()));
     }
-    this.inputPath = ((FileSplit) inputSplit).getPath();
-    this.mappingString = multiInputTaggedSplit.getConf().get(PartitionedFileSetDataset.PATH_TO_PARTITIONING_MAPPING);
+
+    this.conf = multiInputTaggedSplit.getConf();
+    String mappingString = conf.get(PartitionedFileSetDataset.PATH_TO_PARTITIONING_MAPPING);
+    this.pathToPartitionMapping =
+      GSON.fromJson(Objects.requireNonNull(mappingString), STRING_PARTITION_KEY_MAP_TYPE);
```

Comment:
This shouldn't happen since we set it into the conf in `PartitionedFileSetDataset#getInputFormatConfiguration`, so even if we improve the error message, it won't help the user in the case that it does happen (which I'm not sure how it can possibly happen).